src/ejabberd_auth_pki.erl

```diff
@@ -0,0 +1,117 @@
+
+%%-callback authorize(mongoose_credentials:t()) -> {ok, mongoose_credentials:t()} | {error, any()}.
+authorize(Creds) ->
+  case get_username(Creds) of
+    {ok,UserName} ->
+      {ok, mongoose_credentials:extend(Creds,[ {username,    UserName},
+                                               {auth_module, ?MODULE}])};
+    {error, Error} -> {error, Error}
+  end.
+
+
+%%-callback try_register( User :: ejabberd:luser(),
+%%                        Server :: ejabberd:lserver(),
+%%                        Password :: binary()
+%%                      ) -> ok | {error, exists | not_allowed | term()}.
+try_register(_, _, _) -> {error, not_allowed}.
+
+%%-callback dirty_get_registered_users() -> [ejabberd:simple_bare_jid()].
+dirty_get_registered_users() -> [].
+
+%%-callback get_vh_registered_users(Server :: ejabberd:lserver()) -> [ejabberd:simple_bare_jid()].
+get_vh_registered_users(_) -> [].
+
+%%-callback get_vh_registered_users( Server :: ejabberd:lserver(),
+%%                                   Opts :: list()
+%%                                 ) -> [ejabberd:simple_bare_jid()].
+get_vh_registered_users(_, _) -> [].
+
+%%-callback get_vh_registered_users_number(Server :: ejabberd:lserver()) -> integer().
+get_vh_registered_users_number(_) -> 0.
+
+%%-callback get_vh_registered_users_number( Server :: ejabberd:lserver(),
+%%                                          Opts :: list()
+%%                                        ) -> integer().
+get_vh_registered_users_number(_, _) -> 0.
+
+%%-callback get_password( User :: ejabberd:luser(),
+%%                        Server :: ejabberd:lserver()
+%%                      ) -> ejabberd_auth:passterm() | false.
+get_password(_, _) -> false.
+
+%%-callback get_password_s( User :: ejabberd:luser(),
+%%                          Server :: ejabberd:lserver()
+%%                        ) -> binary().
+get_password_s(_, _) -> <<"">>.
+
+%%-callback does_user_exist( User :: ejabberd:luser(),
+%%                           Server :: ejabberd:lserver()
+%%                         ) -> boolean() | {error, atom()}.
+does_user_exist(_, _) -> true.
+
+%%-callback remove_user( User :: ejabberd:luser(),
+%%                       Server :: ejabberd:lserver()
+%%                     ) -> ok | {error, not_allowed}.
+remove_user(_, _) -> {error, not_allowed}.
+
+%%-callback remove_user( User :: ejabberd:luser(),
+%%                       Server :: ejabberd:lserver(),
+%%                       Password :: binary()
+%%                     ) -> ok | {error, not_exists | not_allowed | bad_request}.
+remove_user(_, _, _) -> {error, not_allowed}.
+
+check_password(_, _, _) -> false.
+
+check_password(_, _, _, _, _) -> false.
+
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% internal functions
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+get_username(Cred) ->
+  case mongoose_credentials:get(Cred,common_name,no_cert) of
```

Comment:
whitespaces, probably indention.
We use style

```erlang
    case X of
        something ->
            something
    end.
```

Reply:
please reference to the doc

Reply:
Personally I don't mind lack of newline when all matches in a case are one-liners.